### ISSUE TITLE
UntaggedEnum with bool variant breaks unexpectedly

### ISSUE THREAD

**Issue by davydog187** [rusterlium/rustler#215]
As shown by #214 , if you have an UntaggedEnum with a bool variant, it will break the decoding of struct variants.

```rust
#[derive(NifUntaggedEnum)]
pub enum UntaggedEnum {
    Foo(u32),
    Bar(String),
    Baz(AddStruct),
    Bool(bool), // Only addition
}
```

```bash
  1) test untagged enum transcoder (RustlerTest.CodegenTest)
     test/codegen_test.exs:44
     Assertion with == failed
     code:  assert :invalid_variant == RustlerTest.untagged_enum_echo([1, 2, 3, 4])
     left:  :invalid_variant
     right: true
     stacktrace:
       test/codegen_test.exs:48: (test)
```

I have started looking at `rustler_codegen/src/untagged_enum.rs`, but nothing stuck out to me that would cause this issue, and I'm wondering if this is an issue that needs to be fixed upstream in `syn`.

I would love to help tackle this bug, but looking for guidance on where I should start poking. Thank you

**Comment by evnu** [comment]
The decoder for `bool` uses [truthiness](https://github.com/rusterlium/rustler/blob/master/rustler/src/types/primitive.rs#L57) to determine if the bool is true or false:

```rust
impl<'a> Decoder<'a> for bool {
    fn decode(term: Term<'a>) -> NifResult<bool> {
        Ok(atom::is_truthy(term))
    }
}
```

[is_truthy](https://github.com/rusterlium/rustler/blob/master/rustler/src/types/atom.rs#L121) checks if the term does not equal `:false` or `:nil`:

```rust
pub fn is_truthy(term: Term) -> bool {
    !((term.as_c_arg() == false_().as_c_arg()) || (term.as_c_arg() == nil().as_c_arg()))
}
```

I too find this behaviour surprising and think it should be made explicit. Where truthiness is expected, an newtype `Truthy` could be used:

```rust
struct Truthy(bool);
```

**Comment by evnu** [comment]
Can this be closed as #219 is now merged, or is there more to do here?

**Comment by davydog187** [comment]
Closing!
